doc/statemachines.tex

```diff
@@ -0,0 +1,310 @@
+  An event is a single point in time when the system transitions from one state to an other state.
+  The type system does not represent an intermediate state. It does not model messages that are send at time
+  $t_1$ and arrive later at time $t_2$.
+  
+  Real world systems use message passing. When going from an event-based system (Haskell implementation/ Protocol/Core.hs)
+  
+  to message passing the interpretation/ semantics has to be adapted.
+\item[Timeouts, lost messages, etc]
+  Important aspects are not covered at this level/layer.
+\end{description}
+  
+\subsubsection{Notation} \marginpar{\njd{Question: how formal should it be ?}}
+
+  \marginpar{\njd{This part should be generic for all Protocols that are implemented via the typed-transition framework.}}
+
+  Todo: Definition of the terms used in the document:
+  What are the terms that we use to describe the network layer ?
+  Can we reference a common source / textbook ?
+\begin{description}
+\item[Protocol]
+\item[Node]  
+\item[Agent]
+  A unit of execution that follows particular Protocol.
+  (Could be implemented as a green thread for example.)
+  A Node can run many agents.
+\item[Client] The agent that is active in the initial state.
+\item[Server] The agent that waits for a message clients when it starts.  
+\item[Upstream,Downstream,Topology]
+\end{description}
+
+\begin{description}
+\item[State Machine]
+  The Protocols are described as a state machine.
+  Formally the state machine is a tuple $(S,T,initial,final)...$ (States, Transitions, initial state, finial state)...
+  The specification uses different representations of the state machine that all describe the same protocol
+  but may leave out some aspects.
+  \begin{itemize}
+  \item The state machine can be represented as a collection of tables (States, Transitions/Messages..).
+  There is a one to one correspondence between these tables and the Haskell source.
+\item A transition graph of the global view of the state machine.
+\item A transition graph of the client view.
+\item A transition graph of the server view.
+\end{itemize}
+  
+\item[States]
+  The specification assumes a global view of the state of a system consisting of the client and the server.
+  Client and server are always in identical states (by definition)
+  which also means that client and server simultaneously transit to new states.
+  The global state is not observable by the client nor the server.
+  Possible interpretations: 
+  A sequence of states is a set of points in time that a correct run of the protocol passes through.
+  (todo: draw diagram)
+\item[Messages/Events]
+  A tuple $(lable,data)$. Can be serialized transferred over the network and de-serialized.
+  (authenticity of messages not handled here)
+\item[Transitions]
+  The document describes the global/abstract state transitions of the protocol
```

Comment:
Do you distinguish here the global state (implementation specific state) from the abstract one (implementation agnostic one, encoded by the state machine)?  Maybe it would be nice to define them as such, what do you think?